### PR TITLE
Story/65/alex

### PR DESCRIPTION
**Pull request by alexcouper**
This attacks #65 (finally)

### CODE REVIEW

File: config/local_settings_alex.py [hunk 1]
```diff
@@ -9,3 +9,17 @@
...
+
+JIRA_CONFIG = {
+    'url': '',
+    'user': None,
+    'password': None,
+}
+
+
+try:
+    from non_github_settings import JIRA_CONFIG
```
Comment (txels) [inline config/local_settings_alex.py:22]: :)

File: deploystream/lib/transforms.py [hunk 2]
```diff
@@ -35,18 +35,39 @@ def remap(original, keymap):
...
         API).
+
     :param keymap:
         A dictionary mapping original field name => transformation.
         Transformation can be a single string, representing the new
         name for the field, if only the name changes, or a tuple
         containing `(new_name, value_mapping)`.
+        Original field name can be a single string, representing the location
+        in the original dict to find the value, or a tuple of keys if the value
+        is nested.
```
Comment (txels) [inline deploystream/lib/transforms.py:46]: nice

Comment (txels) [inline deploystream/lib/transforms.py:46]: `:param defaults:` ?

File: deploystream/lib/transforms.py [hunk 3]
```diff
@@ -35,18 +35,39 @@ def remap(original, keymap):
...
+
+    .. note::
+
+        Only keys defined in keymap will be moved across to the new dictionary.
```
Comment (txels) [inline deploystream/lib/transforms.py]: Oh. This is a change in semantics. Why this change?

File: tests/test_lib/test_mappings.py [hunk 4]
```diff
@@ -34,3 +34,15 @@ def test_name_and_value_remap(self):
...
+
+    def test_nested_name_remap(self):
+        """
+        Test nested field names can be mapped.
+        """
+        original = {"somekey": {"identifier": 123, "other_data": 445}}
+        mapping = {('somekey', 'identifier'): "id"}
+        expected = {"id": 123}
+
+        mapped = remap(original, mapping)
```
Comment (txels) [inline tests/test_lib/test_mappings.py]: should include a test case with the additional argument `defaults`.

File: deploystream/providers/jira/__init__.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                - title
+                - id
+                - url
+                - feature_type
+                - owner
+                - description
+
+            ``None`` otherwise
+        """
+        pass
```
Comment (txels) [inline deploystream/providers/jira/__init__.py]: Maybe we should start using `NotImplementedError` instead of `pass` (me guilty)

File: tests/test_providers/test_jira_provider.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    features = jira_provider.get_features()
+
+    assert_equals(features[0]['title'], "Perform update")
+    assert_equals(features[0]['id'], "TD-193")
+    assert_equals(features[0]['url'],
+        "https://something.atlassian.net/rest/api/2/issue/46385")
+    assert_equals(features[0]['type'], "Sub-task")
+    assert_equals(features[0]['owner'], "alex.couper")
+    assert_equals(features[0]['project'], "Superior Project")
```
Comment (txels) [inline tests/test_providers/test_jira_provider.py:39]: of course :)

File: tests/test_providers/test_jira_provider.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from tests import load_fixture
+
+ISSUE_RESPONSE_JSON = load_fixture('jira_issues.json')
+
+
+def test_is_planning_provider():
+    assert_true(is_implementation(JiraProvider, IPlanningProvider))
+
+
+@patch('jira.client.requests')
```
Comment (txels) [inline tests/test_providers/test_jira_provider.py:16]: Ha, this is extreme whitebox, having to look at the source code of an external library in order to write a test case... :P

**Comment by txels** [comment]
Out.

File: deploystream/lib/transforms.py (hunk 3 above)
Comment (alexcouper) [inline deploystream/lib/transforms.py]: We could bring it back if you want. I changed it because JIRA brings back a huge amount of data with each issue so it seemed a "waste" to copy across keys and nested dicts that were never going to be used.

File: deploystream/providers/jira/__init__.py (hunk 5 above)
Comment (alexcouper) [inline deploystream/providers/jira/__init__.py]: Changed

File: tests/test_lib/test_mappings.py (hunk 4 above)
Comment (alexcouper) [inline tests/test_lib/test_mappings.py]: I've removed this additional argument as it seems to have been hangover from an earlier refactor attempt

File: tests/test_providers/test_jira_provider.py (hunk 7 above)
Comment (alexcouper) [inline tests/test_providers/test_jira_provider.py:16]: I know :( - but a prime pretenders refactor candidate.

And it's made me aware of this kind of use-case for pretenders more generally: When using any third party library to connect to an API. 

The only 2 non-pretender ways are to: a) do this extreme whitebox, or b) do similar extreme white box of creating the third party objects and returning them.

File: deploystream/lib/transforms.py (hunk 2 above)
Comment (alexcouper) [inline deploystream/lib/transforms.py:46]: Removed the unused `defaults` param

File: deploystream/lib/transforms.py (hunk 3 above)
Comment (txels) [inline deploystream/lib/transforms.py]: I was thinking that eventually some day the plugins/providers we would build would have a frontend as well as backend bits, and the frontend bits would be able to display additional info that may not be deploystream-normalised data, but provider-specific. There is nothing in this PR that prevents that, but it makes it a little tedious to have to change code in two places for that.

While I was testing Github and sprintly I found it useful to get everything that was coming through, and find additional useful fields that I had not initially considered.

That said, there is nothing wrong with being totally explicit here, but it means I will have to test that github and sprintly still work, which I doubt as I was relying on the fact that some basic fields have typically the same names we expect in deploystream (title, description, etc...).

File: deploystream/providers/jira/__init__.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+from deploystream.lib import transforms
+
+# Map fields from the Jira API to deploystream names
+FEATURE_MAP = {
+    ('fields', 'summary'): 'title',
+    'key': 'id',
+    'self': 'url',
+    ('fields', 'issuetype', 'name'): 'type',
+    ('fields', 'assignee', 'name'): ('owner', {None: ""}),
+    ('fields', 'project', 'name'): 'project',
```
Comment (txels) [inline deploystream/providers/jira/__init__.py:13]: Looks like JIRA returns pretty verbose JSON...

File: deploystream/lib/transforms.py (hunk 3 above)
Comment (txels) [inline deploystream/lib/transforms.py]: Maybe the middle ground is to add a parameter `keep_extra=True` if you want to keep the extra fields that JIRA may invoke with `False`?

File: tests/data/jira_issues.json [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+                        "24x24": "https://something.atlassian.net/secure/useravatar?size=small&ownerId=alex.couper&avatarId=10603",
+                        "192x192": "https://something.atlassian.net/secure/useravatar?size=xxlarge%402x&ownerId=alex.couper&avatarId=10603",
+                        "16x16": "https://something.atlassian.net/secure/useravatar?size=xsmall&ownerId=alex.couper&avatarId=10603",
+                        "96x96": "https://something.atlassian.net/secure/useravatar?size=xxlarge&ownerId=alex.couper&avatarId=10603",
+                        "256x256": "https://something.atlassian.net/secure/useravatar?size=xxxlarge%402x&ownerId=alex.couper&avatarId=10603",
+                        "64x64": "https://something.atlassian.net/secure/useravatar?size=xlarge&ownerId=alex.couper&avatarId=10603",
+                        "128x128": "https://something.atlassian.net/secure/useravatar?size=xxxlarge&ownerId=alex.couper&avatarId=10603",
+                        "32x32": "https://something.atlassian.net/secure/useravatar?size=medium&ownerId=alex.couper&avatarId=10603"
```
Comment (txels) [inline tests/data/jira_issues.json:27]: OMG what a waste of bandwidth

**Comment by alexcouper** [comment]
_cough cough_

**Comment by txels** [comment]
Hi, sorry I've had a shitty week. There is one outstanding comment about the semantic change in remapping. Have you tested that the github provider still works fine? I didn't really have time to test sprintly, may give it a go later.

**Comment by alexcouper** [comment]
Ah yes, you wanted the option to `keep_extra`. I'll add that next time I'm working in this space

**Comment by alexcouper** [comment]
That's sorted now.
